driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionAutoDataKeysTest.java

```diff
@@ -0,0 +1,248 @@
+                .build());
+        client.getDatabase(KEY_VAULT_NAMESPACE.getDatabaseName()).drop();
+        db = client.getDatabase(AbstractClientSideEncryptionAutoDataKeysTest.class.getSimpleName());
+        db.drop();
+    }
+
+    @AfterEach
+    @SuppressWarnings("try")
+    public void cleanUp() {
+        try (ClientEncryption ignored = clientEncryption;
+             MongoClient ignored1 = client) {
+            // empty
+        }
+    }
+
+    /**
+     * See
+     * <a href="https://github.com/mongodb/specifications/tree/master/source/client-side-encryption/tests#case-1-simple-creation-and-validation">
+     * Case 1: Simple Creation and Validation</a>.
+     */
+    @ParameterizedTest(name = DISPLAY_NAME_PLACEHOLDER + " {0}")
+    @MethodSource("arguments")
+    void simpleCreationAndValidation(final KmsProvider kmsProvider) {
+        CreateCollectionOptions createCollectionOptions = new CreateCollectionOptions().encryptedFields(Document.parse(
+                "{"
+                + "  fields: [{"
+                + "    path: 'ssn',"
+                + "    bsonType: 'string',"
+                + "    keyId: null"
+                + "  }]"
+                + "}"));
+        clientEncryption.createEncryptedCollection(db, COLL_NAME, createCollectionOptions,
+                kmsProvider.createEncryptedCollectionParamsSupplier.get());
+        MongoCollection<Document> coll = db.getCollection(COLL_NAME);
+        assertEquals(
+                121, // DocumentValidationFailure
+                assertThrows(MongoWriteException.class, () -> coll.insertOne(Document.parse("{ ssn: '123-45-6789' }")))
+                        .getCode());
+    }
+
+    /**
+     * See
+     * <a href="https://github.com/mongodb/specifications/tree/master/source/client-side-encryption/tests#case-2-missing-encryptedfields">
+     * Case 2: Missing encryptedFields</a>.
+     */
+    @ParameterizedTest(name = DISPLAY_NAME_PLACEHOLDER + " {0}")
+    @MethodSource("arguments")
+    void missingEncryptedFields(final KmsProvider kmsProvider) {
+        assertThrows(MongoConfigurationException.class, () -> clientEncryption.createEncryptedCollection(
+                db, COLL_NAME, new CreateCollectionOptions(), kmsProvider.createEncryptedCollectionParamsSupplier.get()));
+        assertTrue(stream(db.listCollectionNames().spliterator(), false).noneMatch(name -> name.equals(COLL_NAME)));
+    }
+
+    /**
+     * See
+     * <a href="https://github.com/mongodb/specifications/tree/master/source/client-side-encryption/tests#case-3-invalid-keyid">
+     * Case 3: Invalid keyId</a>.
+     */
+    @ParameterizedTest(name = DISPLAY_NAME_PLACEHOLDER + " {0}")
+    @MethodSource("arguments")
+    void invalidKeyId(final KmsProvider kmsProvider) {
+        CreateCollectionOptions createCollectionOptions = new CreateCollectionOptions().encryptedFields(Document.parse(
+                "{"
+                + "  fields: [{"
```

Comment:
Should there be a test for `{fields: []}` ? Is that valid or invalid and does the code create the collection?

Reply:
> Should there be a test for {fields: []} ?

Neither the current spec nor the [upcoming test changes](https://jira.mongodb.org/browse/DRIVERS-2538) have such a test. There are a few existing CSFLE tests that use `fields: []`, for example https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/tests/legacy/fle2-NoEncryption.json#L16-L18.

I don't know if we need a test for this value. Do you want me to add one?

> Is that valid or invalid and does the code create the collection?

Whether an empty `fields` array is a valid or not configuration is not specified anywhere:

- https://www.mongodb.com/docs/manual/reference/command/create/
- https://www.mongodb.com/docs/manual/core/csfle/quick-start/
- https://mongodb.github.io/mongo-java-driver/4.8/apidocs/mongodb-driver-core/com/mongodb/client/model/CreateCollectionOptions.html#encryptedFields(org.bson.conversions.Bson)
- https://mongodb.github.io/mongo-java-driver/4.8/apidocs/mongodb-driver-core/com/mongodb/AutoEncryptionSettings.html#getEncryptedFieldsMap()

An experiment shows that the server creates a collection when `fields` is an empty array.

Reply:
Thanks for looking into it - as CSFLE tests cover it, resolving.